### ISSUE TITLE
Rewrite UI front end using React.js

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#876]
Current UI is functional in certain parts but less functional on others (e.g. Sales and Receiving).
There is still dependency on old style js/jQuery although 3.0.0 has been a huge quantum leap.

React.js looks appealing and thinner for front end UI implementation offering the type of responsiveness we would require at component level.

Also learning React.js could open the door to use react-native to implement some mobile phone app interacting with a future OSPOS Rest API.

We should probably considering this option in the future and see how that matches what we have and not causing a too dramatic rewriting of the whole code, if possible.

Ambitious plans but I sense this could be a good move.

**Comment by jekkos** [comment]
I'm also certain that some of the more complex js stuff would be much simpler using a framework like this.  Most of the pages consist of reusable components anyway (thinking about bootstrap tables in this case) but it will certainly take time to get every detail right again (just comparing with the bstables effort beginning this year)

**Comment by daN4cat** [comment]
To help with any RESTful API development: https://apiary.io/

**Comment by jekkos** [comment]
OK will look into that. We could also use builtin composer support for REST in CI (looks quite easy to setup)

**Comment by daN4cat** [comment]
Yes, we use composer for REST support. I understand that tool helps with API design, documentation and mock up. The documentation part would be handy.

**Comment by jekkos** [comment]
Another option there is [Swagger UI](http://swagger.io/swagger-ui/), which generates documentation, sample web client etc for existing REST API. Very easy to use as well

**Comment by daN4cat** [comment]
Agreed, Swagger is probably widely used.

**Comment by daN4cat** [comment]
I changed the title of this issue because we should evaluate Vue.js Vuex.js (Redux) too.
I heard that Vue.js is a bit more advance than the basic React.js and simplifies quite a lot of the front end work.
However moving to this type of solution implies first a move to REST API and then a switch of FrontEnd technology with quite some work.
It's not a trivial job....

**Comment by jekkos** [comment]
Basically it's more of an application rewrite I think.

**Comment by jekkos** [comment]
Except for the data model which we can keep.

**Comment by daN4cat** [comment]
Yes it is, a Front End application.
The first step would be to work on the REST API and then the rest.
A new set of controllers will deal with the REST API.

**Comment by gaaytrimaa** [comment]
> I changed the title of this issue because we should evaluate Vue.js 2 and Vuex/Redux (Flux) too.
> I heard that Vue.js is a bit more advanced than the basic React.js and simplifies quite a lot of the front end work.
> However moving to this type of solution implies first a move to REST API and then a switch of FrontEnd technology with quite some work.
> It's not a trivial job....

Yes Sir, you are right Vue.js is certainly a great and easy to learn framework to work upon. It would be good investment to start designing the frontend UI in Vue.js. And if we choose Firestore as a backend then it can handle online/offline support. Also, Firestore has various built in user authentication like Gmail, Email, Facebook, Phone authentication, etc.
The most important of all is that Firestore is extremely fast and it would be great if we migrate to it.
Drawbacks:- 
Firestore doesn't have a local development facility so we have to always test the software on a real database on Firebase console.
We will have to design the entire database schema again.

I am ready to work on the frontend if you support me in designing the backend on Firestore.

**Comment by jekkos** [comment]
Ok sounds good. All help is welcome. Never heard about firestore, so I guess I'll have to do my homework there. I think having REST apis in the backend or at least REST specifications for our services would allow for an easy frontend / backend rewrite.

**Comment by crls12opazo** [comment]
I took on the task of rewriting ospos to generate a backend and a frontend, and I accomplished it using C# and .NET Framework, and Ionic with Angular for the front-end. In this way, I maintained the data model of ospos and focused on migrating the code. Also, this made it easier for me to generate an app for both Android and iOS. I believe I have migrated around 60% of the functionalities so far.

![image](https://user-images.githubusercontent.com/32281096/235336645-67755d84-0538-45a0-926b-a65d331db576.png)

![image](https://user-images.githubusercontent.com/32281096/235336669-864bfd3f-e747-4d6d-9217-a25329267ccf.png)

![image](https://user-images.githubusercontent.com/32281096/235336683-9f31a8d9-38e9-482f-9df8-632a1b87b832.png)
